### PR TITLE
Fix `MetricCollection` when input are metrics that return dicts with same keywords

### LINKED ISSUE #2020: MetricCollection NOT working for MeanAveragePrecision
## 🐛 Bug

`MetricCollection` does not work with `MeanAveragePrecision`

### To Reproduce

```py
import torch
from torchmetrics.detection import MeanAveragePrecision
from torchmetrics import MetricCollection

predictions = [{'boxes': torch.tensor([[0, 0, 1, 1]]),
                'labels': torch.tensor([1]),
                'scores': torch.tensor([1.0])}]
targets = [{'boxes': torch.tensor([[0, 0, 1, 1]]),
            'labels': torch.tensor([1])}]

metric = MetricCollection({
    "map30": MeanAveragePrecision(iou_thresholds=[0.3]),
    "map40": MeanAveragePrecision(iou_thresholds=[0.4]),
    "map50": MeanAveragePrecision(iou_thresholds=[0.5]),
})

metric.update(predictions, targets)
metric.compute()

# {'map': tensor(1.),
# 'map_50': tensor(-1.),
# 'map_75': tensor(-1.),
# 'map_small': tensor(1.),
# 'map_medium': tensor(-1.),
# 'map_large': tensor(-1.),
# 'mar_1': tensor(1.),
# 'mar_10': tensor(1.),
# 'mar_100': tensor(1.),
# 'mar_small': tensor(1.),
# 'mar_medium': tensor(-1.),
# 'mar_large': tensor(-1.),
# 'map_per_class': tensor(1.),
# 'mar_100_per_class': tensor(1.),
# 'classes': tensor(1, dtype=torch.int32)}
```

### Expected behavior

````
{'map30' : {'map': ...},
 'map40': {...},
 'map50': {...}
}
````

### Environment

- TorchMetrics version 1.0.1
- Python & PyTorch Version: 3.10.11 and 2.0.1
- Any other relevant information such as OS (e.g., Linux): Linux

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2020 
Currently if metric collection have multiple metrics in the collection that returns dicts and some of the keys in those dicts are the same, it only returns the results of the last metric (because it is overwriting the key). Fixes this.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2027.org.readthedocs.build/en/2027/

### CODE REVIEW

File: src/torchmetrics/collections.py [hunk 1]
```diff
@@ -337,6 +337,8 @@ def _compute_and_reduce(
 
             if isinstance(res, dict):
                 for key, v in res.items():
+                    stripped_k = k.replace(m.prefix, "").replace(m.postfix, "")
```
Comment (Borda) [inline src/torchmetrics/collections.py:340]:
```suggestion
                    stripped_k = k.replace(getattr(m, "prefix", ""), "")
                    stripped_k = k.replace(getattr(m, "postfix", ""), "")
```

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
